app_test.go

```diff
@@ -2265,3 +2265,239 @@ func TestWhenExitSubCommandWithCodeThenAppQuitUnexpectedly(t *testing.T) {
 		t.Errorf("exitCodeFromOsExiter valeu should be %v, but its value is %v", testCode, exitCodeFromExitErrHandler)
 	}
 }
+
+func TestDuplicateCommandName(t *testing.T) {
+	duplicateCommandName := "hello"
+	os.Args = []string{"say", duplicateCommandName}
+	app := NewApp()
+	app.Name = "say"
+	app.Commands = []Command{
+		{
+			Name: duplicateCommandName,
+			Action: func(c *Context) error {
+				fmt.Println("Hello")
+				return nil
+			},
+		},
+		{
+			Name: duplicateCommandName,
+			Action: func(c *Context) error {
+				fmt.Println("Hello duplicate")
+				return nil
+			},
+		},
+	}
+
+	err := app.Run(os.Args)
+	if err == nil {
+		t.Fatalf("expected to receive error from Run, got none")
+	}
+
+	if err.Error() != fmt.Sprintf("Your app contains multiple commands with the Name %q. Having multiple commands with the same name results in ambiguous behavior, so please make sure each command in your app has a unique name.", duplicateCommandName) {
+		t.Errorf("received error did not match expectation")
+	}
+}
+
+func TestDuplicateSubcommandName(t *testing.T) {
+	duplicateSubcommandName := "word"
+	os.Args = []string{"say", "kind", duplicateSubcommandName}
+	app := NewApp()
+	app.Name = "say"
+	app.Commands = []Command{
+		{
+			Name: "kind",
+			Subcommands: []Command{
+				{
+					Name: duplicateSubcommandName,
+					Action: func(c *Context) error {
+						fmt.Println("please")
+						return nil
+					},
+				},
+				{
+					Name: duplicateSubcommandName,
+					Action: func(c *Context) error {
+						fmt.Println("thanks")
+						return nil
+					},
+				},
+			},
+		},
+	}
+
+	err := app.Run(os.Args)
+	if err == nil {
+		t.Fatalf("expected to receive error from Run, got none")
+	}
+
+	if err.Error() != fmt.Sprintf("Your command %q contains multiple subcommands with the Name %q. Having multiple subcommands with the same name results in ambiguous behavior, so please make sure each subcommand in your command has a unique name.", "kind", duplicateSubcommandName) {
+		t.Errorf("received error did not match expectation")
+	}
+}
+
+func TestDuplicateSubcommandNameOnSubcommandLevel(t *testing.T) {
```

Comment:
⭐⭐⭐ _(must change)_ We need another test case here, specifically a test case where the commands look like this:

```
- top-level-command
  - subcommand-a
    - subcommand-1
  - subcommand-b
    - subcommand-1
```

Where we assert that if there is an error message, it isn't our "duplicate commands" error message.